Mathlib/Tactic/HashCommandLinter.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Std.Lean.Command
+import Std.Data.Array.Basic
+
+/-!
+#  `#`-command linter
+
+The `#`-command linter produces a warning when a command starting with `#` is used.
+-/
+
+namespace Mathlib.Linter
+
+/--
+Setting to `true` enables the '`#`-command' linter.  When `true`, Lean warns on any command
+beginning with `#`.  For example, `#eval, #guard, #guard_msgs` all trigger a message.
+-/
+register_option linter.hashCommand : Bool := {
+  defValue := false
+  descr := "enable the `#`-command linter"
+}
+
+namespace HashCommandLinter
+
+open Lean Elab
+
+/-- `getAtomStx stx` extracts the array of all `.atom` nodes in `stx`. -/
+def getAtomStx : Syntax → Array Syntax
+  | .node _ _ args => (args.attach.map fun ⟨a, _h⟩ => getAtomStx a).foldl (· ++ ·) default
+  | s@(.atom ..) => #[s]
+  | _ => default
+
+/-- Gets the value of the `linter.hashCommand` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.hashCommand o
+
+/-- `whiteList` is the array of `#`-commands that are allowed in 'Mathlib'. -/
+private abbrev whiteList : Array String := #["#align", "#align_import", "#noalign"]
+
+/-- Checks that no command beginning with `#` is present in 'Mathlib', except for the ones in
+`whiteList`. -/
+def hashCommandLinter : Linter where run := withSetOptionIn fun stx => do
```

Comment:
We should also discard syntax like `variable ... in` and `open ... in` before linting. We can do that with something like the following, which is mostly cargo-culted from `withSetOptionIn`, but splits on the nodekind of the thing before `in`. (Not tested in this file, since my computer is slow and switching branches in VS code right now is a hassle! :P Definitely deserves tests.)

```suggestion

open Command in
/-- Exactly like `withSetOptionIn`, but recursively discards any other use of `in`. -/
private partial def withSetOptionIn' (cmd : CommandElab) : CommandElab := fun stx => do
  if stx.getKind == ``Lean.Parser.Command.in then
    if stx[0].getKind == ``Lean.Parser.Command.set_option then
      let opts ← Elab.elabSetOption stx[0][1] stx[0][2]
      Command.withScope (fun scope => { scope with opts }) do
        withSetOptionIn' cmd stx[1]
    else
      withSetOptionIn' cmd stx[1]
  else
    cmd stx

/-- Checks that no command beginning with `#` is present in 'Mathlib', except for the ones in
`whiteList`. -/
def hashCommandLinter : Linter where run := withSetOptionIn' fun stx => do
```

Reply:
I have not been able to get this to work.  As is, I think that it does not really enter the various `in`s.  Also, `#guard_msgs` should really be linter against anyway.

I wonder if I should follow more closely the string-based linter:
* extract the commands that begin with `#` from the environment;
* scan the syntax of the command to be linted for nodes that match the `#`-commands;
* warn on a match found.

Reply:
Many apologies for the delay here—I didn't see this comment and thought you were still digesting it! 😅 (And for the past few days I've been a bit busy.)

There's a good reason my suggestion didn't work: it's wrong! :) I cargo-culted from `withSetOptionIn`, and there's a bug in `withSetOptionIn`. (leanprover/lean4#3806) Namely, it should recurse into `stx[2]` (the command syntax), not `stx[1]` (which is the token `"in"`). You'll notice that currently,
```lean
set_option lean.hashCommand true in
#print 0
```
doesn't work as expected.

So it should actually be 
```lean
open Command in
/-- Exactly like `withSetOptionIn`, but recursively discards any other use of `in`. Intended to be 
used in the hashcommand linter, where we don't need to handle non-`set_option` `in` commands. -/
private partial def withSetOptionIn' (cmd : CommandElab) : CommandElab := fun stx => do
  if stx.getKind == ``Lean.Parser.Command.in then
    if stx[0].getKind == ``Lean.Parser.Command.set_option then
      let opts ← Elab.elabSetOption stx[0][1] stx[0][2]
      Command.withScope (fun scope => { scope with opts }) do
        withSetOptionIn' cmd stx[2]
    else
      withSetOptionIn' cmd stx[2]
  else
    cmd stx
```
Note also that since `#guard_msgs in` doesn't actually use the `Lean.Parser.Command.in` parser for `in` (its `" in"` is defined in its own syntax, unlike e.g. `set_option ... in`), it's exempt from `withSetOptionIn'`. (This is also why the linter doesn't fire twice in each `#guard_msgs` test: at the top level, it hits the docstring on `#guard_msgs` and stops, which is actually quite useful.)

A good test for `withSetOptionIn'` would be something like
```lean
variable [Nonempty x] in
#synth Nonempty x
```

Since the only reason this linter doesn't catch `#guard_msgs` is the presence of a preceding docstring, and since I'm pretty sure that's unique to `#guard_msgs`, I'm okay with `#guard_msgs` not being linted against! Linting against `#guard_msgs` would also make testing the linter itself difficult. Especially since an accidental `#guard_msgs` would be (I think!) very rare.

I'm happy to push these changes if it would be convenient for you (I've now tested them locally), but if you'd like to make them yourself that's fine too! :)